File: frontend/src/lib/api/ad-library.ts
/**
 * Ad Library prospecting API client.
 *
 * Pull advertisers from public ad libraries (Meta / Google), surface the ones
 * running consistently but not iterating creatives, and ingest qualified
 * advertisers into the CRM. Types are sourced from the generated OpenAPI spec
 * so they stay in lockstep with the backend.
 */

import { apiDelete, apiGet, apiPatch, apiPost } from "@/lib/api";
import { queryKeys } from "@/lib/query-keys";
import { POLL_30S } from "@/lib/query-options";

import type { Schemas } from "./_client";

export type AdLibrarySearchRequest = Schemas["AdLibrarySearchRequest"];
export type AdLibraryJob = Schemas["LeadDiscoveryJobResponse"];
export type AdAdvertiser = Schemas["AdAdvertiserResponse"];
export type AdAdvertiserDetail = Schemas["AdAdvertiserDetail"];
export type AdCreative = Schemas["AdCreativeResponse"];
export type AdSignalBreakdown = Schemas["AdSignalBreakdown"];
export type PaginatedAdAdvertisers = Schemas["PaginatedAdAdvertisers"];
export type AdMonitor = Schemas["AdMonitorResponse"];
export type AdMonitorCreate = Schemas["AdMonitorCreate"];
export type AdMonitorUpdate = Schemas["AdMonitorUpdate"];
export type AdvertiserPromoteRequest = Schemas["AdvertiserPromoteRequest"];
export type AdvertiserPromoteResult = Schemas["AdvertiserPromoteResult"];
export type AdvertiserBulkPromoteRequest = Schemas["AdvertiserBulkPromoteRequest"];
export type AdvertiserBulkPromoteResult = Schemas["AdvertiserBulkPromoteResult"];
export type IcpThresholds = Schemas["IcpThresholds"];

export interface ListAdvertisersParams {
  platform?: "meta" | "google";
  only_qualified?: boolean;
  contact_traced?: boolean;
  page?: number;
  page_size?: number;
}

function advertisersQueryString(params: ListAdvertisersParams): string {
  const search = new URLSearchParams();
  if (params.platform) search.set("platform", params.platform);
  if (params.only_qualified !== undefined)
    search.set("only_qualified", String(params.only_qualified));
  if (params.contact_traced !== undefined)
    search.set("contact_traced", String(params.contact_traced));
  if (params.page !== undefined) search.set("page", String(params.page));
  if (params.page_size !== undefined) search.set("page_size", String(params.page_size));
  const query = search.toString();
  return query ? `?${query}` : "";
}

export const adLibraryApi = {
  search: (workspaceId: string, request: AdLibrarySearchRequest): Promise<AdLibraryJob> =>
    apiPost<AdLibraryJob>(`/api/v1/workspaces/${workspaceId}/ad-library/search`, request),

  getJob: (workspaceId: string, jobId: string): Promise<AdLibraryJob> =>
    apiGet<AdLibraryJob>(`/api/v1/workspaces/${workspaceId}/ad-library/jobs/${jobId}`),

  listAdvertisers: (
    workspaceId: string,
    params: ListAdvertisersParams = {},
  ): Promise<PaginatedAdAdvertisers> =>
    apiGet<PaginatedAdAdvertisers>(
      `/api/v1/workspaces/${workspaceId}/ad-library/advertisers${advertisersQueryString(params)}`,
    ),

  getAdvertiser: (workspaceId: string, advertiserId: string): Promise<AdAdvertiserDetail> =>
    apiGet<AdAdvertiserDetail>(
      `/api/v1/workspaces/${workspaceId}/ad-library/advertisers/${advertiserId}`,
    ),

  promoteAdvertiser: (
    workspaceId: string,
    advertiserId: string,
    request: AdvertiserPromoteRequest,
  ): Promise<AdvertiserPromoteResult> =>
    apiPost<AdvertiserPromoteResult>(
      `/api/v1/workspaces/${workspaceId}/ad-library/advertisers/${advertiserId}/promote`,
      request,
    ),

  bulkPromote: (
    workspaceId: string,
    request: AdvertiserBulkPromoteRequest,
  ): Promise<AdvertiserBulkPromoteResult> =>
    apiPost<AdvertiserBulkPromoteResult>(
      `/api/v1/workspaces/${workspaceId}/ad-library/advertisers/bulk-promote`,
      request,
    ),

  listMonitors: (workspaceId: string): Promise<AdMonitor[]> =>
    apiGet<AdMonitor[]>(`/api/v1/workspaces/${workspaceId}/ad-library/monitors`),

  createMonitor: (workspaceId: string, request: AdMonitorCreate): Promise<AdMonitor> =>
    apiPost<AdMonitor>(`/api/v1/workspaces/${workspaceId}/ad-library/monitors`, request),

  updateMonitor: (
    workspaceId: string,
    monitorId: string,
    request: AdMonitorUpdate,
  ): Promise<AdMonitor> =>
    apiPatch<AdMonitor>(
      `/api/v1/workspaces/${workspaceId}/ad-library/monitors/${monitorId}`,
      request,
    ),

  deleteMonitor: (workspaceId: string, monitorId: string): Promise<void> =>
    apiDelete(`/api/v1/workspaces/${workspaceId}/ad-library/monitors/${monitorId}`),
};

/** React Query option presets for ad-library reads. */
export const adLibraryQueryOptions = {
  advertisers: (workspaceId: string, params: ListAdvertisersParams = {}) => ({
    queryKey: queryKeys.adLibrary.advertisers(workspaceId, { ...params }),
    queryFn: () => adLibraryApi.listAdvertisers(workspaceId, params),
  }),
  advertiser: (workspaceId: string, advertiserId: string) => ({
    queryKey: queryKeys.adLibrary.advertiser(workspaceId, advertiserId),
    queryFn: () => adLibraryApi.getAdvertiser(workspaceId, advertiserId),
  }),
  monitors: (workspaceId: string) => ({
    queryKey: queryKeys.adLibrary.monitors(workspaceId),
    queryFn: () => adLibraryApi.listMonitors(workspaceId),
  }),
  /** Poll a discovery job while it runs. */
  job: (workspaceId: string, jobId: string) => ({
    queryKey: queryKeys.adLibrary.job(workspaceId, jobId),
    queryFn: () => adLibraryApi.getJob(workspaceId, jobId),
    ...POLL_30S,
  }),
};

File: backend/app/api/v1/ad_library.py
"""Ad Library prospecting API endpoints.

Pull advertisers from public ad libraries (Meta Ad Library, Google Ads
Transparency), surface the ones running consistently but **not** iterating
creatives, and ingest qualified advertisers into the CRM. The
:class:`AdLibraryService` owns workspace scoping; this router keeps request
validation + dependency wiring at the boundary.
"""

from __future__ import annotations

import uuid

from fastapi import APIRouter, Query, status

from app.api.deps import DB, CurrentUser, WorkspaceAccess
from app.api.service_errors import ServiceErrorRoute
from app.models.ad_advertiser import AdPlatform
from app.schemas.ad_advertiser import AdAdvertiserDetail, PaginatedAdAdvertisers
from app.schemas.ad_library import (
    AdLibrarySearchRequest,
    AdMonitorCreate,
    AdMonitorResponse,
    AdMonitorUpdate,
    AdvertiserBulkPromoteRequest,
    AdvertiserBulkPromoteResult,
    AdvertiserPromoteRequest,
    AdvertiserPromoteResult,
)
from app.schemas.lead_discovery_job import LeadDiscoveryJobResponse
from app.services.ad_intelligence.ad_library_service import AdLibraryService

router = APIRouter(route_class=ServiceErrorRoute)


@router.post(
    "/search",
    response_model=LeadDiscoveryJobResponse,
    status_code=status.HTTP_202_ACCEPTED,
)
async def create_search(
    workspace_id: uuid.UUID,
    search_in: AdLibrarySearchRequest,
    current_user: CurrentUser,
    db: DB,
    workspace: WorkspaceAccess,
) -> LeadDiscoveryJobResponse:
    """Launch an ad-library search.

    Enqueues a discovery job the ad-library worker runs asynchronously; poll
    ``GET /jobs/{id}`` for status.
    """
    service = AdLibraryService(db)
    job = await service.create_search_job(workspace_id, search_in, requested_by_id=current_user.id)
    return LeadDiscoveryJobResponse.model_validate(job)


@router.get("/jobs/{job_id}", response_model=LeadDiscoveryJobResponse)
async def get_job(
    workspace_id: uuid.UUID,
    job_id: uuid.UUID,
    current_user: CurrentUser,
    db: DB,
    workspace: WorkspaceAccess,
) -> LeadDiscoveryJobResponse:
    """Return the status of an ad-library discovery job."""
    service = AdLibraryService(db)
    job = await service.get_job(workspace_id, job_id)
    return LeadDiscoveryJobResponse.model_validate(job)


@router.get("/advertisers", response_model=PaginatedAdAdvertisers)
async def list_advertisers(
    workspace_id: uuid.UUID,
    current_user: CurrentUser,
    db: DB,
    workspace: WorkspaceAccess,
    platform: AdPlatform | None = Query(default=None),
    only_qualified: bool = Query(
        default=False,
        description="Apply ICP floors + exclude prolific testers.",
    ),
    contact_traced: bool | None = Query(default=None),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=25, ge=1, le=100),
) -> PaginatedAdAdvertisers:
    """List tracked advertisers ranked by opportunity score (ICP fit)."""
    service = AdLibraryService(db)
    return await service.list_advertisers(
        workspace_id,
        platform=platform,
        only_qualified=only_qualified,
        contact_traced=contact_traced,
        page=page,
        page_size=page_size,
    )


@router.get("/advertisers/{advertiser_id}", response_model=AdAdvertiserDetail)
async def get_advertiser(
    workspace_id: uuid.UUID,
    advertiser_id: uuid.UUID,
    current_user: CurrentUser,
    db: DB,
    workspace: WorkspaceAccess,
) -> AdAdvertiserDetail:
    """Return one advertiser with creatives, signal breakdown, and traced contact."""
    service = AdLibraryService(db)
    return await service.get_advertiser_detail(workspace_id, advertiser_id)


@router.post(
    "/advertisers/{advertiser_id}/promote",
    response_model=AdvertiserPromoteResult,
)
async def promote_advertiser(
    workspace_id: uuid.UUID,
    advertiser_id: uuid.UUID,
    promote_in: AdvertiserPromoteRequest,
    current_user: CurrentUser,
    db: DB,
    workspace: WorkspaceAccess,
) -> AdvertiserPromoteResult:
    """Promote one advertiser into the CRM (advertiser -> prospect -> contact)."""
    service = AdLibraryService(db)
    return await service.promote_advertiser(workspace_id, advertiser_id, promote_in)


@router.post("/advertisers/bulk-promote", response_model=AdvertiserBulkPromoteResult)
async def bulk_promote_advertisers(
    workspace_id: uuid.UUID,
    promote_in: AdvertiserBulkPromoteRequest,
    current_user: CurrentUser,
    db: DB,
    workspace: WorkspaceAccess,
) -> AdvertiserBulkPromoteResult:
    """Promote many advertisers into the CRM in one call."""
    service = AdLibraryService(db)
    return await service.bulk_promote(workspace_id, promote_in)


@router.get("/monitors", response_model=list[AdMonitorResponse])
async def list_monitors(
    workspace_id: uuid.UUID,
    current_user: CurrentUser,
    db: DB,
    workspace: WorkspaceAccess,
) -> list[AdMonitorResponse]:
    """List saved ad-library monitors (recurring ICP searches)."""
    service = AdLibraryService(db)
    return await service.list_monitors(workspace_id)


@router.post(
    "/monitors",
    response_model=AdMonitorResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_monitor(
    workspace_id: uuid.UUID,
    monitor_in: AdMonitorCreate,
    current_user: CurrentUser,
    db: DB,
    workspace: WorkspaceAccess,
) -> AdMonitorResponse:
    """Create a saved monitor that re-scans on a schedule."""
    service = AdLibraryService(db)
    return await service.create_monitor(workspace_id, monitor_in, created_by_id=current_user.id)


@router.patch("/monitors/{monitor_id}", response_model=AdMonitorResponse)
async def update_monitor(
    workspace_id: uuid.UUID,
    monitor_id: uuid.UUID,
    monitor_in: AdMonitorUpdate,
    current_user: CurrentUser,
    db: DB,
    workspace: WorkspaceAccess,
) -> AdMonitorResponse:
    """Update a saved monitor's search / thresholds / schedule / active flag."""
    service = AdLibraryService(db)
    return await service.update_monitor(workspace_id, monitor_id, monitor_in)


@router.delete("/monitors/{monitor_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_monitor(
    workspace_id: uuid.UUID,
    monitor_id: uuid.UUID,
    current_user: CurrentUser,
    db: DB,
    workspace: WorkspaceAccess,
) -> None:
    """Delete a saved monitor."""
    service = AdLibraryService(db)
    await service.delete_monitor(workspace_id, monitor_id)
